oauthbearer.go

```diff
@@ -61,3 +64,121 @@ func (a *oauthBearerClient) Next(challenge []byte) ([]byte, error) {
 func NewOAuthBearerClient(opt *OAuthBearerOptions) Client {
 	return &oauthBearerClient{*opt}
 }
+
+type OAuthBearerAuthenticator func(opts OAuthBearerOptions) *OAuthBearerError
+
+type oauthBearerServer struct {
+	done         bool
+	fail         error
+	authenticate OAuthBearerAuthenticator
+}
+
+func (a *oauthBearerServer) Next(response []byte) (challenge []byte, done bool, err error) {
+	// Per RFC, we cannot just send an error, we need to return JSON-structured
+	// value as a challenge and then after getting dummy response from the
+	// client stop the exchange.
+	if a.fail != nil {
+		if len(response) != 1 && response[0] != 0x01 {
+			return nil, true, errors.New("unexpected response")
+		}
+		return nil, true, a.fail
+	}
+	fail := func(descr string) ([]byte, bool, error) {
+		blob, err := json.Marshal(OAuthBearerError{
+			Status:  "invalid_request",
+			Schemes: "bearer",
+		})
+		if err != nil {
+			panic(err) // wtf
+		}
+		a.fail = errors.New(descr)
+		return blob, false, nil
+	}
+
+	if a.done {
+		err = ErrUnexpectedClientResponse
+		return
```

Comment:
These should be avoided

https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters

Reply:
(The rest of the codebase has the wrong style because it's old)

Reply:
You probably want to clean plain.go too then :)
https://github.com/emersion/go-sasl/blob/6fe53deab1f6dadb88217928f90c11179a5885df/plain.go#L47-L48